language-plutus-core/src/Language/PlutusCore/Type.hs

```diff
@@ -91,29 +93,57 @@ instance Corecursive (Type tyname a) where
     embed (TyLamF l tn k ty)    = TyLam l tn k ty
     embed (TyAppF l ty ty')     = TyApp l ty ty'
 
--- | Substitute @a@ for @b@ in @c@.
-tyNameSubstitute :: Eq (tyname a)
-                 => tyname a -- ^ @a@
-                 -> tyname a -- ^ @b@
-                 -> Type tyname a -- ^ @c@
-                 -> Type tyname a
-tyNameSubstitute tn tn' = cata a where
-    a (TyVarF x tn'') | tn == tn'' = TyVar x tn'
-    a x               = embed x
-
-instance (Eq (tyname a), Eq a) => Eq (Type tyname a) where
-    (==) (TyVar a1 tn) (TyVar a2 tn')                   = a1 == a2 && tn == tn'
-    (==) (TyFun a1 ty ty') (TyFun a2 ty'' ty''')        = a1 == a2 && ty == ty'' && ty' == ty'''
-    (==) (TyFix a1 tn ty) (TyFix a2 tn' ty')            =
-        a1 == a2 && tyNameSubstitute tn' tn ty' == ty
-    (==) (TyForall a1 tn k ty) (TyForall a2 tn' k' ty') =
-        a1 == a2 && tyNameSubstitute tn' tn ty' == ty && k == k'
-    (==) (TyBuiltin a1 b) (TyBuiltin a2 b')             = a1 == a2 && b == b'
-    (==) (TyInt a1 n) (TyInt a2 n')                     = a1 == a2 && n == n'
-    (==) (TyLam a1 tn k ty) (TyLam a2 tn' k' ty')       =
-        a1 == a2 && tyNameSubstitute tn' tn ty' == ty && k == k'
-    (==) (TyApp a1 ty ty') (TyApp a2 ty'' ty''')        = a1 == a2 && ty == ty'' && ty' == ty'''
-    (==) _ _                                            = False
+-- this type is used for replacing type names in
+-- the Eq instance
+type EqState tyname a = M.Map (tyname a) (tyname a)
+
+rebind :: (Ord (tyname a), MonadState (EqState tyname a) m)
+       => tyname a
+       -> tyname a
+       -> m ()
+rebind = modify .* M.insert
+
+rebindAndEq :: (Eq a, Ord (tyname a), MonadState (EqState tyname a) m)
+            => Type tyname a
+            -> Type tyname a
+            -> tyname a
+            -> tyname a
+            -> m Bool
+rebindAndEq ty ty' tn tn' =
+    rebind tn' tn *>
+    eqTypeM ty ty'
+
+-- This tests for equality of names inside a monad that allows substitution.
+eqTypeM :: (Ord (tyname a), MonadState (EqState tyname a) m, Eq a)
+        => Type tyname a
+        -> Type tyname a
+        -> m Bool
+
+eqTypeM (TyFun _ ty ty') (TyFun _ ty'' ty''') = (&&) <$> eqTypeM ty ty'' <*> eqTypeM ty' ty'''
+eqTypeM (TyApp _ ty ty') (TyApp _ ty'' ty''') = (&&) <$> eqTypeM ty ty'' <*> eqTypeM ty' ty'''
+
+eqTypeM (TyInt _ n) (TyInt _ n')              = pure (n == n')
+eqTypeM (TyBuiltin _ b) (TyBuiltin _ b')      = pure (b == b')
+
+eqTypeM (TyFix _ tn ty) (TyFix _ tn' ty') =
+    rebindAndEq ty ty' tn tn'
+eqTypeM (TyForall _ tn k ty) (TyForall _ tn' k' ty') = do
+    tyEq <- rebindAndEq ty ty' tn tn'
+    pure (tyEq && k == k')
```

Comment:
The `k == k'` check is simpler than `rebindAndEq ty ty' tn tn'`, so I think it should run first.

Reply:
Still remains. I'd write this as `pure (k == k') <&&> rebindAndEq ty ty' tn tn'` modulo new naming.

Reply:
Or we can define something like

```haskell
(&&>) :: Monad m => Bool -> m Bool -> m Bool
```

and use it.

The reason for using either this combinator or `(<&&>)` together with `pure` is that if kinds do not match, you do not need to traverse any other parts of the types you're comparing for equality. But right now you always recurse if `m` is a strict monad. Which it's not I guess, because it's instantiated to the lazy `State`, but

1. the type signature has `MonadState`, i.e. no guarantees that this will be turned into lazy `State`
2. relying on laziness like that is a bad idea, at some point you'll just trick yourself. It's better to treat all monads as strict unless there is a specific reason not to do so. This code can be easily rewritten into something that makes sense both strictly and lazily, so I do not see such reason here

Reply:
> Still remains.

What I commented, was actually fixed. Sorry, again being unclear.